File: book/petrol-pump/chapter3/section5/ClearSalePump.java
package chapter3.section5;

import pump.*;
import chapter3.section2.LifeCycle;
import chapter3.section4.Fill;
import chapter3.section4.ShowDollarsPump;
import sodium.*;
import java.util.Optional;

public class ClearSalePump implements Pump {
    public Outputs create(Inputs inputs) {
        EventLoop<Fuel> eStart = new EventLoop<>();
        Fill fi = new Fill(
                          inputs.eClearSale.map(u -> Unit.UNIT),
                          inputs.eFuelPulses, inputs.calibration,
                          inputs.price1, inputs.price2, inputs.price3,
                          eStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.eNozzle1,
                              inputs.eNozzle2,
                              inputs.eNozzle3),
                inputs.eClearSale,
                fi);
        eStart.loop(np.eStart);
        return new Outputs()
            .setDelivery(np.fuelFlowing.map(
                of ->
                    of.equals(Optional.of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.equals(Optional.of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.equals(Optional.of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .setSaleCostLCD(fi.dollarsDelivered.map(
                    q -> Formatters.formatSaleCost(q)))
            .setSaleQuantityLCD(fi.litersDelivered.map(
                    q -> Formatters.formatSaleQuantity(q)))
            .setPriceLCD1(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.ONE, inputs))
            .setPriceLCD2(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.TWO, inputs))
            .setPriceLCD3(ShowDollarsPump.priceLCD(np.fillActive, fi.price,
                    Fuel.THREE, inputs))
            .setBeep(np.eBeep)
            .setSaleComplete(np.eSaleComplete);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section7/ClearSalePump.cs
using PetrolPump.Chapter3.Section3;
using PetrolPump.Chapter3.Section6;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section7
{
  public class ClearSalePump : IPump 
  {
    public Outputs Create(Inputs inputs) 
    {
        EventLoop<Fuel> eStart = new EventLoop<Fuel>();
        Fill fi = new Fill(
                          inputs.EClearSale.Map(u => Unit.UNIT),
                          inputs.EFuelPulses, inputs.Calibration,
                          inputs.Price1, inputs.Price2, inputs.Price3,
                          eStart);
        NotifyPointOfSale np = new NotifyPointOfSale(
                new LifeCycle(inputs.ENozzle1,
                              inputs.ENozzle2,
                              inputs.ENozzle3),
                inputs.EClearSale,
                fi);
        eStart.Loop(np.EStart);
        return new Outputs()
            .SetDelivery(np.FuelFlowing.Map(
                of =>
                    of.Equals(Optional<Fuel>.Of(Fuel.ONE))   ? Delivery.FAST1 :
                    of.Equals(Optional<Fuel>.Of(Fuel.TWO))   ? Delivery.FAST2 :
                    of.Equals(Optional<Fuel>.Of(Fuel.THREE)) ? Delivery.FAST3 : 
                                                         Delivery.OFF))
            .SetSaleCostLcd(fi.DollarsDelivered.Map(
                    q => Formatters.FormatSaleCost(q)))
            .SetSaleQuantityLcd(fi.LitersDelivered.Map(
                    q => Formatters.FormatSaleQuantity(q)))
            .SetPriceLcd1(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price,
                    Fuel.ONE, inputs))
            .SetPriceLcd2(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price,
                    Fuel.TWO, inputs))
            .SetPriceLcd3(ShowDollarsPump.PriceLCD(np.FillActive, fi.Price,
                    Fuel.THREE, inputs))
            .SetBeep(np.EBeep)
            .SetSaleComplete(np.ESaleComplete);
    }
  }
}
